os/arch/arm/src/armv7-a/arm_mmu.c

```diff
@@ -254,3 +281,190 @@ void mmu_invalidate_region(uint32_t vstart, size_t size)
+ * Name: mmu_allocate_app_l2_pgtbl
+ *
+ * Description:
+ *   Allocate space for L2 page table of application, in accordance with
+ *   the requirements of the arch specific mmu.
+ *
+ * Input Parameters:
+ *
+ * Returned Value:
+ * L2 Page table address
+ ****************************************************************************/
+uint32_t *mmu_allocate_app_l2_pgtbl(int app_id, int l2_idx)
+{
+	uint32_t *addr = (uint32_t *)(PGTABLE_BASE_VADDR + (3 * 16384) + (app_id * 4096) + (l2_idx * 1024));
+	return addr;
+}
+
+/****************************************************************************
+ * Name: mmu_update_app_l1_pgtbl_ospgtbl
+ *
+ * Description:
+ * Loop through the L1 page table.
+ * Copy kernel L1 page table to app page table.
+ * If the entry is pointing to a L2 page table
+ * Allocate L2 page table for app.
+ * Copy entries from kernel to app L2 table.
+ * Update the L2 page table address in L1 table.
+ * 
+ * Input Parameters:
+ * app_pgtbl: Pointer to L1 page table of app
+ *
+ ****************************************************************************/
+void mmu_update_app_l1_pgtbl_ospgtbl(uint32_t *app_l1_pgtbl)
+{
+	uint32_t *os_l1_pgtbl = (uint32_t *)PGTABLE_BASE_VADDR;
+	memcpy((void *)app_l1_pgtbl, (void *)os_l1_pgtbl, L1_PGTBL_SIZE);
+}
+
+/****************************************************************************
+ * Name: mmu_map_app_region
+ *
+ * Description
+ *
+ * Input Parameters:
+ *
+ * Returned Value:
+ ****************************************************************************/
+void mmu_map_app_region(int app_id, uint32_t *l1_pgtbl, uint32_t start, uint32_t size, bool ro, bool exec, bool global)
+{
+	uint32_t idx;
+	uint32_t val;
+	uint32_t end = start + size;
+	int l2_idx = 0;
+
+	// Run a loop until the entire region is mapped.
+	while (start < end) {
+		// Check if this address can be mapped to a section.
+		if (!(start & SECTION_MASK) && !(size & SECTION_MASK)) {
+			// Yes. Update the section entry in the the L1 page table.
+			idx = start >> 20;
+			val = start & PMD_PTE_PADDR_MASK;
+			mmu_set_flags(&val, ro, exec, true);
+			l1_pgtbl[idx] = val;
+
+			// Advance the memory region address.
+			start += SECTION_SIZE;
+		} else {	// Check if this address can be mapped to a small page.
+
+			// Check if L2 page table is not created.
+			idx = (start & 0xfff00000) >> 20;
+			uint32_t *l2_pgtbl = (uint32_t *)(l1_pgtbl[idx] & PMD_PTE_PADDR_MASK);
+			
+			if ((l1_pgtbl[idx] & PMD_TYPE_MASK) != PMD_TYPE_PTE) {
+				// Yes. Allocate L2 page table for app.
+				l2_pgtbl = mmu_allocate_app_l2_pgtbl(app_id, l2_idx++);
+
+				// Fill the newly allocated l2 page table with default kernel page entries
+				uint32_t l2_start = start & PMD_SECT_PADDR_MASK;
+                                for (int i = 0; i < L2_PGTBL_NENTRIES; i++) {
+                                        l2_pgtbl[i] = l2_start | MMU_L2_MEMFLAGS;
+                                        l2_start += 0x1000;
```

Comment:
It's better to use ```SMALL_PAGE_SIZE``` here instead of 0x1000.

Reply:
Done.